docs/winrt/WinRT SxS Components.md

```diff
@@ -1 +1,153 @@
+
+```
+namespace Widgets_1.Contoso.Widgets {
+        runtime_class Doodad;
+}
+```
+
+Version 2 of the same component will end up with a different implicit namespace in metadata:
+
+```
+namespace Widgets_1.Contoso.Widgets {
+        runtime_class Doodad;
+}
+```
+
+### ABI & Platform support
+At the ABI, types will self-identify using the component prefix namespace. For example, the IInspectable::GetRuntimeClassName method for Doodad will return the string "Widgets_1.Contoso.Widgets.Doodad". This ensure that metadata resolution algorithms and metadata-based marshalling support in all versions of Windows that support it will be able to uniquely correlate runtime classes with metadata definitions, without risk of type confusion.
+
+The component-qualified type name will also be used in calculations for pinterfaces GUID calculations. This ensures that marshaling and dynamic casting via QueryInterface will not result in type confusion.
+
+Component authors that use explicit GUIDs for interfaces are encouraged, but not required to generate new GUIDs if the binary layout of an interface changes across major versions. 
+
+Type registrations will include the component namespace element in the registered type name. This ensures that activtion of different versions is unambiguous, and that an application can include more than one version of the component. Call to RoGetActivationFactory or RoActivateInstance should use the component-qualified name>
+
+### The C++ Developer Experience
+Developers using older versions of C++ /WinRT will see the qualified namespace name. Developers using enlightened versions of C++ /WinRT will use the components as if the component namespace does not exist. This ensures that only minimal work is required to migrate from version 1 to version 2 of a component. Migration work can focus on breaking changes across versions, rather than mechanical changes required to update references to version-specific names.
+
+Open Design Issues: Should it be a goal to allow a single C++ /WinRT consumer to work with two versions of a component within the same compilation unit?
+
+### The C# Developer Experience
+Clients using .NET Core 3.1 and earlier, as well as .NET Framework 4 will be able to use components, but will refer to types using component-qualified names.
+
+Clients using the .NET 5 projection will not use component-qualified names. Instead, the projections will be differentiated by assembly name.
+
+#### Consuming
+
+Open Design Issue: How do the projection tools and build system handle uniquely named assemblies for different versions of the same component?
+
+#### Authoring
+
+Open Design Issue: How do the projection tools and build system handle uniquely named assemblies for different versions of the same component? Can the single source file authoring experience provide a high-quality experience for implementing components?
```

Comment:
Also for authoring, I _think_ your proposal would let C++ developers produce a single DLL that implements multiple versions of the component, which would be pretty keen from a back/forward compat POV.

I could completely break my interface in v2.0 of the component, but still have a single type that effectively implements both the v1.5 and v2.0 interfaces.

Reply:
It depends a bit on the details of how C++ allocates "good names". if they're aliased through namespaces and sufficient knobs are provided, it's possible that a single binary or possibly even a single compilation unit could use multiple versions.

Reply:
Hmm, this could best be achieved by the component author by enabling the versioned namespace in the projection. C# for .NET 5 would have some challenges with this, I think, because of an optimization for .NET -> .NET calls, but C++ and perhaps Rust could handle this gracefully, as could the older .NET projection, which would still just see these as different types.